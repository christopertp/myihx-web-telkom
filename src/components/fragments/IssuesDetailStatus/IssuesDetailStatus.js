import React from 'react';
import moment from 'moment';
import { useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';
import Avatar from '../../elements/Avatar';
import Bullet from '../../elements/Bullet';
import Button from '../../elements/Button';
import Card from '../../elements/Card';
import Heading from '../../elements/Heading';
import IconButton from '../../elements/IconButton';
import styles from './styles.scoped.css';

const loading = <span className="loading" />;

export default function IssuesDetailStatus() {
  const { type } = useSelector(s => s.issues);
  const isLogic = type === 'logic';
  return (
    <>
      {!isLogic && <Appointment />}
      <Ticket />
      {!isLogic && <Address />}
      <Technician />
      <Steps />
    </>
  );
}

export function Appointment() {
  const { isLoading, schedule } = useSelector(s => s.issues);
  const { date, slot } = schedule;
  return (
    <section className={styles.appointment}>
      <Heading>Your Appointment</Heading>
      <Card variant="hover">
        <figure>
          {
            (!isLoading.d && date) ?
              (<>
                <small>{moment(date).format('ddd')}</small>
                <p>{moment(date).format('D')}</p>
                <small>{moment(date).format('MMMM')}</small>
              </>) : ''
          }
        </figure>
        <h4>Repair Request</h4>
        {isLoading.d ? loading : <p>Estimated arrival time {slot}</p>}
        <Button variant="text">Reschedule</Button>
      </Card>
    </section>
  );
}

export function Ticket() {
  const { isLoading, summary, type } = useSelector(s => s.issues);
  const { id } = useParams();
  const isLogic = type === 'logic';
  return (
    <section className={styles.ticket}>
      <Heading>Your {isLogic ? 'Report' : 'Ticket'}</Heading>
      <Card className={styles[type]} variant="hover">
        {(!isLoading.d && isLogic) && <h4>{id}</h4>}
        {isLoading.d ? loading : summary}
      </Card>
    </section>
  );
}

export function Address() {
  const { address, isLoading } = useSelector(s => s.issues);
  return (
    <section className={styles.address}>
      <Heading>Address</Heading>
      {isLoading.d ? loading : <p>{address}</p>}
    </section>
  );
}

export function Technician() {
  const { technician } = useSelector(s => s.issues);
  const { name, status, tel } = technician;

  if (!name) {
    return <Heading>Report Process</Heading>;
  }

  const onCall = url => () => {
    location.href = url;
  };

  return (
    <section className={styles.technician}>
      <Heading>Your Technician</Heading>
      <Card variant="hover">
        <Avatar src="/assets/av_blank.svg" />
        <h4>{name}</h4>
        <p>{status}</p>
        <IconButton name="chat" onClick={onCall(`https://api.whatsapp.com/send?phone=${tel}`)} />
        <IconButton name="call" onClick={onCall(`tel:${tel}`)} />
      </Card>
    </section>
  );
}

export function Steps() {
  const { step, type } = useSelector(s => s.issues);
  const STEPS = {
    fisik: [
      {
        title: 'Ticket Submitted',
        desc: 'Ticket has been submitted. Our team is currently processing your request',
      },
      {
        title: 'Ticket Assigned',
        desc: 'We have assigned a technician to your request',
      },
      {
        title: 'Repair In Progress',
        desc: 'Our technician is arriving to your location',
      },
      {
        title: 'Repair Complete',
        desc: 'Issue has been resolved!',
      },
    ],
    logic: [
      {
        title: 'Report Received',
        desc: 'We are trying to solve this issue',
      },
      {
        title: 'Repair In Progress',
        desc: 'We have assigned a special team to resolve this issue',
      },
      {
        title: 'Repair Complete',
        desc: 'Our team has finished repairing the issue!',
      },
      {
        title: 'Issue Solved',
        desc: 'Issue has been resolved!',
      },
    ],
  };
  const GREEN = '#0eb37e';
  const GREY = '#e5e6e6';

  const setColor = idx => {
    if (idx > step) {
      return [GREY, GREY];
    } else if (idx === step) {
      return [GREEN, GREY];
    }
    return [GREEN, GREEN];
  };

  return (
    <ul className={styles.steps}>
      {STEPS[type]?.map((i, idx) => (
        <li className={idx > step ? styles['step-inactive'] : ''} key={idx}>
          <Bullet color={setColor(idx)} connect={idx < STEPS[type].length - 1} />
          <h4>{i.title}</h4>
          <p>{i.desc}</p>
        </li>
      ))}
    </ul>
  );
}
