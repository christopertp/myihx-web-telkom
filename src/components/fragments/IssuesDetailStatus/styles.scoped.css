.address, .appointment, .technician, .ticket {
  margin-bottom: 1.5rem;
}

.appointment > div > h4,
.steps > li > h4,
.technician > div > h4 {
  color: #191A1A;
  font-size: 1.125rem;
  font-weight: 800;
}

.appointment > div > p,
.steps > li > p,
.technician > div > p {
  color: #66686A;
  font-size: 0.9rem;
  line-height: 1.4;
}

.address > p {
  color: #333333;
  line-height: 1.4;
  margin-top: 0.5rem;
}

.address > span {
  display: block;
  height: 1.25rem;
  margin-top: 0.5rem;
  width: 60%;
}

.appointment > div {
  display: grid;
  grid-template-columns: auto 1fr;
  grid-template-rows: auto auto auto;
  margin-top: 0.5rem;
}

.appointment > div > figure {
  background: linear-gradient(45deg, #333435, #808285);
  border-radius: 0.5rem;
  color: #ffffff;
  display: flex;
  flex-direction: column;
  grid-row: 1/4;
  height: 5.25rem;
  justify-content: center;
  margin-right: 1rem;
  text-align: center;
  width: 5.25rem;
}

.appointment > div > figure > p {
  font-size: 2rem;
  font-weight: 800;
  line-height: 1;
}
.appointment > div > figure > small { font-size: 0.8rem; }
.appointment > div > figure > small:last-child {
  opacity: 0.5;
  text-transform: uppercase;
}

.appointment > div > p { margin: 0.25rem 0 0.5rem; }

.appointment > div > span {
  height: 1.25rem;
  margin: 0.25rem 0 0.5rem;
  width: 13.75rem;
}

.appointment > div > button {
  font-size: 1rem;
  margin-right: auto;
  padding: 0 !important;
  text-transform: capitalize !important;
}

.steps { margin-top: 0.5rem; }

.steps > li {
  display: grid;
  grid-template-columns: auto 1fr;
  grid-template-rows: auto auto;
}

.steps > li > div {
  grid-row: 1/3;
  margin-right: 1rem;
  padding-top: 0.5rem;
}

.steps > li > p { margin: 0.25rem 0 2rem; }

.steps > .step-inactive > h4 {
  color: #b2b4b5;
  font-weight: 400;
}

.steps > .step-inactive > p {
  color: #b2b4b5;
}

.technician > div {
  display: grid;
  grid-template-columns: auto 1fr auto auto;
  grid-template-rows: auto auto;
  margin-top: 0.5rem;
}

.technician > div > figure {
  grid-row: 1/3;
  height: 3.125rem;
  margin-right: 1rem;
  width: 3.125rem;
}

.technician > div > h4 {
  grid-column: 2;
  grid-row: 1;
}

.technician > div > p {
  grid-row: 2;
  grid-column: 2;
}

.technician > div > button {
  grid-row: 1/3;
  margin: auto auto auto 1rem;
}

.ticket > div {
  color: #333333;
  line-height: 1.4;
  margin-top: 0.5rem;
}

.ticket > div > h4 {
  color: #191a1a;
  font-size: 1.125rem;
}

.ticket > div > span {
  display: block;
  height: 1.25rem;
  width: 60%;
}

@media only screen and (max-width: 48em) {
  .appointment > div > h4,
  .steps > li > h4,
  .technician > div > h4 {
    font-size: 1rem;
  }

  .appointment > div > p,
  .steps > li > p,
  .technician > div > p {
    font-size: 0.75rem;
  }

  .address { margin-bottom: 1.625rem; }
  .address > p { font-size: 0.9rem; }

  .appointment { margin-bottom: 1.5rem; }
  .appointment > div > figure {
    height: 3.75rem;
    margin-right: 0.75rem;
    width: 3.75rem;
  }
  .appointment > div > figure > p { font-size: 1.5rem; }
  .appointment > div > figure > small { font-size: 0.75rem; }
  .appointment > div > p { margin: 0.25rem 0; }
  .appointment > div > span {
    height: 1rem;
    margin: 0.25rem 0;
    width: 80%;
  }  
  .appointment > div > button { font-size: 0.8rem; }

  .steps { padding: 0 1rem; }
  .steps > li > div {
    padding-top: 0.25rem;
    margin-right: 0.8rem;
  }
  .steps > li > p { margin: 0.25rem 0 1rem; }

  .technician { margin-bottom: 1rem; }
  .technician > div > figure {
    height: 2.5rem;
    width: 2.5rem;
  }
  .technician > div > p { margin-top: 0.25rem; }

  .ticket { margin-bottom: 2rem; }
  .ticket > .fisik {
    box-shadow: none;
    font-size: 0.9rem;
    padding: 0;
  }
  .ticket > .logic > h4 { font-size: 1rem; }
}
