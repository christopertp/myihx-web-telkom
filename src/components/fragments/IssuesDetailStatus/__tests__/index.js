import React from 'react';
import { useSelector } from 'react-redux';
import ShallowRenderer from 'react-test-renderer/shallow';
import IssuesDetailStatus, { Appointment, Ticket, Address, Technician, Steps } from '../IssuesDetailStatus';

const store = {
  address: 'address',
  schedule: {
    date: '2011-10-01',
    slot: '10.00',
  },
  isLoading: {
    d: false,
  },
  message: '',
  step: 2,
  summary: 'Issue',
  technician: {
    name: 'Telkom',
    tel: '08123456',
  },
  type: 'fisik',
};

useSelector.mockImplementation(fn => {
  fn({});
  return store;
});

describe('src/components/fragments/IssuesDetailStatus', () => {
  test('render', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<IssuesDetailStatus />);
    expect(tree).toMatchSnapshot();
  });

  test('Appointment', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Appointment />);
    expect(tree).toMatchSnapshot();
  });

  test('Appointment loading', () => {
    useSelector.mockImplementationOnce(() => ({ ...store, isLoading: { d: true } }));
    const res = Appointment();
    const card = res.props.children[1];
    expect(card.props.children[0].props.children).toBe('');
    expect(card.props.children[2].props.className).toBe('loading');
  });

  test('Ticket', () => {
    useSelector.mockImplementationOnce(fn => {
      fn({});
      return { ...store, type: 'logic' };
    });
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Ticket />);
    expect(tree).toMatchSnapshot();
  });

  test('Ticket loading', () => {
    useSelector.mockImplementationOnce(() => ({ ...store, isLoading: { d: true } }));
    const res = Ticket();
    expect(res.props.children[1].props.children[1].props.className).toBe('loading');
  });

  test('Address', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Address />);
    expect(tree).toMatchSnapshot();
  });

  test('Address loading', () => {
    useSelector.mockImplementationOnce(() => ({ ...store, isLoading: { d: true } }));
    const res = Address();
    expect(res.props.children[1].props.className).toBe('loading');
  });

  test('Technician', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Technician />);
    expect(tree).toMatchSnapshot();
  });

  test('Technician onCall & state hidden', () => {
    delete window.location;
    window.location = {
      href: '',
    };
    const res1 = Technician();
    res1.props.children[1].props.children[4].props.onClick();
    expect(window.location.href).toBe('tel:08123456');

    useSelector.mockImplementationOnce(() => ({ ...store, technician: { name: '' } }));
    const res2 = Technician();
    expect(res2.props.children).toBe('Report Process');
  });

  test('Steps', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Steps />);
    expect(tree).toMatchSnapshot();
  });
});
