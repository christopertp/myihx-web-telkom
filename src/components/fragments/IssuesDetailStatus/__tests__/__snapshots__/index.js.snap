// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/fragments/IssuesDetailStatus Address 1`] = `
<section>
  <Heading
    className=""
  >
    Address
  </Heading>
  <p>
    address
  </p>
</section>
`;

exports[`src/components/fragments/IssuesDetailStatus Appointment 1`] = `
<section>
  <Heading
    className=""
  >
    Your Appointment
  </Heading>
  <Card
    className=""
    disabled={false}
    onClick={null}
    to=""
    variant="hover"
  >
    <figure>
      <React.Fragment>
        <small>
          ddd
        </small>
        <p>
          D
        </p>
        <small>
          MMMM
        </small>
      </React.Fragment>
    </figure>
    <h4>
      Repair Request
    </h4>
    <p>
      Estimated arrival time 
      10.00
    </p>
    <Button
      className=""
      disabled={false}
      onClick={null}
      to=""
      type="button"
      variant="text"
    >
      Reschedule
    </Button>
  </Card>
</section>
`;

exports[`src/components/fragments/IssuesDetailStatus Steps 1`] = `
<ul>
  <li
    className=""
  >
    <Bullet
      className=""
      color={
        Array [
          "#0eb37e",
          "#0eb37e",
        ]
      }
      connect={true}
      size="0.75rem"
    />
    <h4>
      Ticket Submitted
    </h4>
    <p>
      Ticket has been submitted. Our team is currently processing your request
    </p>
  </li>
  <li
    className=""
  >
    <Bullet
      className=""
      color={
        Array [
          "#0eb37e",
          "#0eb37e",
        ]
      }
      connect={true}
      size="0.75rem"
    />
    <h4>
      Ticket Assigned
    </h4>
    <p>
      We have assigned a technician to your request
    </p>
  </li>
  <li
    className=""
  >
    <Bullet
      className=""
      color={
        Array [
          "#0eb37e",
          "#e5e6e6",
        ]
      }
      connect={true}
      size="0.75rem"
    />
    <h4>
      Repair In Progress
    </h4>
    <p>
      Our technician is arriving to your location
    </p>
  </li>
  <li>
    <Bullet
      className=""
      color={
        Array [
          "#e5e6e6",
          "#e5e6e6",
        ]
      }
      connect={false}
      size="0.75rem"
    />
    <h4>
      Repair Complete
    </h4>
    <p>
      Issue has been resolved!
    </p>
  </li>
</ul>
`;

exports[`src/components/fragments/IssuesDetailStatus Technician 1`] = `
<section>
  <Heading
    className=""
  >
    Your Technician
  </Heading>
  <Card
    className=""
    disabled={false}
    onClick={null}
    to=""
    variant="hover"
  >
    <Avatar
      className=""
      src="/assets/av_blank.svg"
    />
    <h4>
      Telkom
    </h4>
    <p />
    <IconButton
      className=""
      name="chat"
      onClick={[Function]}
      title=""
      to=""
      type="button"
    />
    <IconButton
      className=""
      name="call"
      onClick={[Function]}
      title=""
      to=""
      type="button"
    />
  </Card>
</section>
`;

exports[`src/components/fragments/IssuesDetailStatus Ticket 1`] = `
<section>
  <Heading
    className=""
  >
    Your 
    Report
  </Heading>
  <Card
    className=""
    disabled={false}
    onClick={null}
    to=""
    variant="hover"
  >
    <h4 />
    Issue
  </Card>
</section>
`;

exports[`src/components/fragments/IssuesDetailStatus render 1`] = `
<React.Fragment>
  <Appointment />
  <Ticket />
  <Address />
  <Technician />
  <Steps />
</React.Fragment>
`;
